File: jfingerprintmanager/src/main/java/com/jesusm/jfingerprintmanager/encryption/EncryptionManager.java
package com.jesusm.jfingerprintmanager.encryption;

import android.support.annotation.NonNull;
import android.support.v4.app.FragmentManager;
import android.support.v4.hardware.fingerprint.FingerprintManagerCompat;

import com.jesusm.jfingerprintmanager.JFingerprintManager;
import com.jesusm.jfingerprintmanager.base.BaseFingerprintManager;
import com.jesusm.jfingerprintmanager.base.FingerprintAssetsManager;
import com.jesusm.jfingerprintmanager.base.ui.System;
import com.jesusm.jfingerprintmanager.encryption.ui.FingerprintEncryptionDialogFragment;
import com.jesusm.jfingerprintmanager.utils.TextUtils;

import java.io.UnsupportedEncodingException;
import java.security.spec.InvalidParameterSpecException;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.spec.IvParameterSpec;

public class EncryptionManager extends BaseFingerprintManager{

    private Encoder encoder;
    private final TextUtils textUtils;

    public EncryptionManager(FingerprintAssetsManager fingerprintAssetsManager, System system,
                             Encoder encoder) {
        super(fingerprintAssetsManager, system);
        this.encoder = encoder;
        this.textUtils = new TextUtils();
    }

    public void encrypt(@NonNull final String messageToEncrypt,
                        @NonNull final JFingerprintManager.EncryptionCallback callback,
                        @NonNull final FragmentManager fragmentManager) {
        if (textUtils.isEmpty(messageToEncrypt)) {
            callback.onEncryptionFailed();
            return;
        }

        fingerprintAssetsManager.initSecureDependencies(new JFingerprintManager.InitialisationCallback() {
            @Override
            public void onErrorFingerprintNotInitialised() {
                callback.onEncryptionFailed();
            }

            @Override
            public void onErrorFingerprintNotEnrolled() {
                callback.onFingerprintNotAvailable();
            }

            @Override
            public void onInitialisationSuccessfullyCompleted() {
                JFingerprintManager.EncryptionAuthenticatedCallback encryptionCallback = new JFingerprintManager.EncryptionAuthenticatedCallback() {
                    @Override
                    public void onAuthenticationSuccess(@NonNull FingerprintManagerCompat.CryptoObject cryptoObject) {
                        Cipher cipher = cryptoObject.getCipher();
                        try {
                            byte[] messageToEncryptBytes = cipher.doFinal(messageToEncrypt.getBytes("UTF-8"));
                            byte[] ivBytes = cipher.getParameters().getParameterSpec(IvParameterSpec.class).getIV();

                            EncryptionData encryptedMessage = new EncryptionData(messageToEncryptBytes, ivBytes, encoder);
                            callback.onEncryptionSuccess(encryptedMessage.printEncryptedInformation());
                        } catch (UnsupportedEncodingException | InvalidParameterSpecException |
                                BadPaddingException | IllegalBlockSizeException e) {
                            callback.onEncryptionFailed();
                        }
                    }

                    @Override
                    public void onFingerprintNotRecognized() {
                        callback.onFingerprintNotRecognized();
                    }

                    @Override
                    public void onAuthenticationFailedWithHelp(String help) {
                        callback.onAuthenticationFailedWithHelp(help);
                    }

                    @Override
                    public void onFingerprintNotAvailable() {
                        callback.onFingerprintNotAvailable();
                    }

                };

                FingerprintEncryptionDialogFragment.Builder builder = new FingerprintEncryptionDialogFragment.Builder();
                showFingerprintDialog(builder, fragmentManager, encryptionCallback);
            }

            @Override
            public void onFingerprintNotAvailable() {
                callback.onFingerprintNotAvailable();
            }
        });
    }

    public void decrypt(@NonNull final String messageToDecrypt,
                        @NonNull final JFingerprintManager.DecryptionCallback callback,
                        @NonNull final FragmentManager fragmentManager) {
        if (textUtils.isEmpty(messageToDecrypt)) {
            callback.onDecryptionFailed();
            return;
        }

        EncryptionData decryptionData = new EncryptionData(messageToDecrypt, encoder);

        if (!decryptionData.dataIsCorrect()) {
            callback.onDecryptionFailed();
            return;
        }

        fingerprintAssetsManager.initSecureDependenciesForDecryption(new JFingerprintManager.InitialisationCallback() {
            @Override
            public void onErrorFingerprintNotInitialised() {
                callback.onDecryptionFailed();
            }

            @Override
            public void onErrorFingerprintNotEnrolled() {
                callback.onFingerprintNotAvailable();
            }

            @Override
            public void onInitialisationSuccessfullyCompleted() {
                JFingerprintManager.EncryptionAuthenticatedCallback decryptionCallback = new JFingerprintManager.EncryptionAuthenticatedCallback() {
                    @Override
                    public void onAuthenticationSuccess(@NonNull FingerprintManagerCompat.CryptoObject cryptoObject) {
                        try {
                            EncryptionData decryptionData = new EncryptionData(messageToDecrypt, encoder);
                            Cipher cipher = cryptoObject.getCipher();

                            byte[] encryptedMessage = decryptionData.message();
                            byte[] decryptedMessageBytes = cipher.doFinal(encryptedMessage);
                            String decryptedMessage = new String(decryptedMessageBytes);

                            callback.onDecryptionSuccess(decryptedMessage);
                        } catch (IllegalBlockSizeException | BadPaddingException e) {
                            callback.onDecryptionFailed();
                        }
                    }

                    @Override
                    public void onFingerprintNotRecognized() {
                        callback.onFingerprintNotRecognized();
                    }

                    @Override
                    public void onAuthenticationFailedWithHelp(String help) {
                        callback.onAuthenticationFailedWithHelp(help);
                    }

                    @Override
                    public void onFingerprintNotAvailable() {
                        callback.onFingerprintNotAvailable();
                    }

                };

                FingerprintEncryptionDialogFragment.Builder builder = new FingerprintEncryptionDialogFragment.Builder();
                showFingerprintDialog(builder, fragmentManager, decryptionCallback);
            }

            @Override
            public void onFingerprintNotAvailable() {
                callback.onFingerprintNotAvailable();
            }
        }, decryptionData.getIVs());
    }
}

File: jfingerprintmanager/src/main/kotlin/com/jesusm/jfingerprintmanager/encryption/EncryptionManager.kt
package com.jesusm.jfingerprintmanager.encryption

import android.support.v4.app.FragmentManager
import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
import com.jesusm.jfingerprintmanager.JFingerprintManager
import com.jesusm.jfingerprintmanager.base.BaseFingerprintManager
import com.jesusm.jfingerprintmanager.base.FingerprintAssetsManager
import com.jesusm.jfingerprintmanager.base.ui.System
import com.jesusm.jfingerprintmanager.encryption.ui.FingerprintEncryptionDialogFragment
import java.io.UnsupportedEncodingException
import java.security.spec.InvalidParameterSpecException
import javax.crypto.BadPaddingException
import javax.crypto.IllegalBlockSizeException
import javax.crypto.spec.IvParameterSpec

class EncryptionManager(val encoder: Encoder, fingerprintAssetsManager: FingerprintAssetsManager, system: System) : BaseFingerprintManager(fingerprintAssetsManager, system) {

    fun encrypt(messageToEncrypt: String, encryptionCallback: JFingerprintManager.EncryptionCallback, fragmentManager: FragmentManager) {
        if (messageToEncrypt.isNullOrEmpty()) {
            encryptionCallback.onEncryptionFailed()
            return
        }

        fingerprintAssetsManager.initSecureDependencies(
                object : JFingerprintManager.InitialisationCallback {
                    override fun onErrorFingerprintNotInitialised() {
                        encryptionCallback.onEncryptionFailed()
                    }

                    override fun onErrorFingerprintNotEnrolled() {
                        encryptionCallback.onFingerprintNotAvailable()
                    }

                    override fun onInitialisationSuccessfullyCompleted() {
                        val callback = object : JFingerprintManager.EncryptionAuthenticatedCallback {
                            override fun onAuthenticationSuccess(cryptoObject: FingerprintManagerCompat.CryptoObject) {
                                val cipher = cryptoObject.cipher
                                try {
                                    val messageToEncryptBytes = cipher.doFinal(messageToEncrypt.toByteArray(charset("UTF-8")))
                                    val ivBytes = cipher.parameters.getParameterSpec(IvParameterSpec::class.java).iv

                                    val encryptedMessage = EncryptionData(messageToEncryptBytes, ivBytes, encoder)
                                    encryptionCallback.onEncryptionSuccess(encryptedMessage.print())
                                } catch (e: UnsupportedEncodingException) {
                                    encryptionCallback.onEncryptionFailed()
                                } catch (e: InvalidParameterSpecException) {
                                    encryptionCallback.onEncryptionFailed()
                                } catch (e: BadPaddingException) {
                                    encryptionCallback.onEncryptionFailed()
                                } catch (e: IllegalBlockSizeException) {
                                    encryptionCallback.onEncryptionFailed()
                                }

                            }

                            override fun onFingerprintNotRecognized() {
                                encryptionCallback.onFingerprintNotRecognized()
                            }

                            override fun onAuthenticationFailedWithHelp(help: String?) {
                                encryptionCallback.onAuthenticationFailedWithHelp(help)
                            }

                            override fun onFingerprintNotAvailable() {
                                encryptionCallback.onFingerprintNotAvailable()
                            }

                            override fun onCancelled() {
                                encryptionCallback.onCancelled()
                            }
                        }

                        val builder = FingerprintEncryptionDialogFragment.Builder()
                        showFingerprintDialog(builder, fragmentManager, callback)
                    }

                    override fun onFingerprintNotAvailable() {
                        encryptionCallback.onFingerprintNotAvailable()
                    }
                })
    }

    fun decrypt(messageToDecrypt: String, callback: JFingerprintManager.DecryptionCallback, fragmentManager: FragmentManager) {
        if (messageToDecrypt.isNullOrEmpty()) {
            callback.onDecryptionFailed()
            return
        }

        val decryptionData = EncryptionData(messageToDecrypt, encoder)
        if (decryptionData.dataIsCorrect().not()) {
            callback.onDecryptionFailed()
            return
        }

        fingerprintAssetsManager.initSecureDependenciesForDecryption(object : JFingerprintManager.InitialisationCallback {
            override fun onErrorFingerprintNotInitialised() {
                callback.onDecryptionFailed()
            }

            override fun onErrorFingerprintNotEnrolled() {
                callback.onFingerprintNotAvailable()
            }

            override fun onInitialisationSuccessfullyCompleted() {
                val decryptionCallback = object : JFingerprintManager.EncryptionAuthenticatedCallback {
                    override fun onAuthenticationSuccess(cryptoObject: FingerprintManagerCompat.CryptoObject) {
                        try {
                            val cipher = cryptoObject.cipher

                            val encryptedMessage = decryptionData.decodedMessage()
                            val decryptedMessageBytes = cipher.doFinal(encryptedMessage)
                            val decryptedMessage = String(decryptedMessageBytes)

                            callback.onDecryptionSuccess(decryptedMessage)
                        } catch (e: IllegalBlockSizeException) {
                            callback.onDecryptionFailed()
                        } catch (e: BadPaddingException) {
                            callback.onDecryptionFailed()
                        }
                    }

                    override fun onFingerprintNotRecognized() {
                        callback.onFingerprintNotRecognized()
                    }

                    override fun onAuthenticationFailedWithHelp(help: String?) {
                        callback.onAuthenticationFailedWithHelp(help)
                    }

                    override fun onFingerprintNotAvailable() {
                        callback.onFingerprintNotAvailable()
                    }

                    override fun onCancelled() {
                        callback.onCancelled()
                    }
                }

                val builder = FingerprintEncryptionDialogFragment.Builder()
                showFingerprintDialog(builder, fragmentManager, decryptionCallback)
            }

            override fun onFingerprintNotAvailable() {
                callback.onFingerprintNotAvailable()
            }
        }, decryptionData.decodedIVs())
    }
}
